### DISCUSSION TITLE
Is the DuckDB query parser exposed to the outside world?

### DISCUSSION THREAD

**Discussion by jivanic-demystdata** [duckdb/duckdb#14962] [Q&A]
We use DuckDB to execute some queries that our users enter in our UI. 
We don't allow all the possible types of queries (only "SELECT" kind of queries), and we want to be able to validate the syntax of the query as soon as possible to provide nice feedback to our users.
For that, we feel that using the parser from DuckDB would be the best solution to be sure that the query is valid

Is it possible to call this parser? 
We're on top of the JVM. Is it already exposed in the Java driver?
If not, we can probably call the C function from Java is this function is exposed/accessible 🤔

Cc @guizmaii

**Comment by Maxxen** [comment]
Look into the `json_serialize_sql` function in the json extension.

**Comment by jivanic-demystdata** [comment]
Thanks for the pointer @Maxxen. Is this function only parsing `SELECT` queries? 🤔

**Reply by Maxxen** [reply]
Yes

**Reply by jivanic-demystdata** [reply]
Is there something else to parse more complex queries?

Dumb query example I'd like to parse could be something like:
```sql
CREATE table ...;
INSERT INTO ...;
SELECT ...;
```

(The "query" is composed of the 3 "statements")

**Comment by jack-w-davis** [comment]
@jivanic-demystdata did you ever find a solution to this? I'm looking to do the same thing

**Comment by Dtenwolde** [comment]
Are you looking only to validate the query (i.e. no parser error) or return an AST as well? It's a bit of a workaround, which we could improve upon, but you might be able to use the `check_peg_parser` table function in the `autocomplete` extension:
```sql
memory D CALL check_peg_parser($TEST_PEG_PARSER$CREATE TABLE database.schema.table(col INTEGER);$TEST_PEG_PARSER$);
┌─────────┐
│ success │
│ boolean │
└─────────┘
  0 rows 
memory D CALL check_peg_parser($TEST_PEG_PARSER$CREATE TAB database.schema.table(col INTEGER);$TEST_PEG_PARSER$);
Binder Error:
Failed to parse query "CREATE TAB database.schema.table(col INTEGER);" - did not consume all tokens (got to token 0 - CREATE)
Tokens:
 0:CREATE
 1:TAB
 2:database
 3:.
 4:schema
 5:.
 6:table
 7:(
 8:col
 9:INTEGER
10:)
11:;
```

We can probably think of additional table functions to provide more information on the parsed query :)

cc @jack-w-davis
